cglib/src/main/java/net/sf/cglib/proxy/Enhancer.java

```diff
@@ -413,10 +394,58 @@ private void validate() {
 
     static class EnhancerFactoryData {
         public final Class generatedClass;
-        public volatile Factory factory;
+        private final Method setThreadCallbacks;
+        private final Class[] primaryConstructorArgTypes;
+        private final Constructor primaryConstructor;
 
-        public EnhancerFactoryData(Class generatedClass) {
+        public EnhancerFactoryData(Class generatedClass, Class[] primaryConstructorArgTypes, boolean classOnly) {
             this.generatedClass = generatedClass;
+            try {
+                setThreadCallbacks = getCallbacksSetter(generatedClass, SET_THREAD_CALLBACKS_NAME);
+                setThreadCallbacks.setAccessible(true);
+                if (classOnly) {
+                    this.primaryConstructorArgTypes = null;
+                    this.primaryConstructor = null;
+                } else {
+                    this.primaryConstructorArgTypes = primaryConstructorArgTypes;
+                    this.primaryConstructor = generatedClass.getDeclaredConstructor(primaryConstructorArgTypes);
+                    primaryConstructor.setAccessible(true);
+                }
+            } catch (NoSuchMethodException e) {
+                throw new CodeGenerationException(e);
+            }
+        }
+
+        public Object newInstance(Class[] argumentTypes, Object[] arguments, Callback[] callbacks) {
+            setThreadCallbacks(callbacks);
+            try {
+                if (!Arrays.equals(primaryConstructorArgTypes, argumentTypes)) {
+                    return ReflectUtils.newInstance(generatedClass, argumentTypes, arguments);
+                }
+                try {
+                    return primaryConstructor.newInstance(arguments);
+                } catch (InstantiationException e) {
+                    throw new CodeGenerationException(e);
+                } catch (IllegalAccessException e) {
+                    throw new CodeGenerationException(e);
+                } catch (InvocationTargetException e) {
+                    throw new CodeGenerationException(e.getTargetException());
+                }
+            } finally {
+                // clear thread callbacks to allow them to be gc'd
+                setThreadCallbacks(null);
+            }
+
+        }
+
+        private void setThreadCallbacks(Callback[] callbacks) {
```

Comment:
Yes, lots of things within cglib are confusing and if designed from scratch might not turn out as they did.  Oh well, we live with what we have.

I understand why you added this code, for precisely the reasons you outline.  My question is: are the old methods still necessary, and if so, please add documentation explaining when you'd use one vs the other.  The documentation is intended for the reader of the code, to understand WTF they're seeing, not for the user of Enhancer.